core/src/OpenACC/Kokkos_OpenACC.cpp

```diff
@@ -73,8 +74,23 @@ Kokkos::Experimental::OpenACC::OpenACC(int async_arg)
 
 void Kokkos::Experimental::OpenACC::impl_initialize(
     InitializationSettings const& settings) {
-  int device_id                   = Kokkos::Impl::get_gpu(settings);
-  Impl::OpenACCInternal::m_accDev = device_id;
+  if (Impl::OpenACC_Traits::may_fallback_to_host &&
+      acc_get_num_devices(Impl::OpenACC_Traits::dev_type) == 0 &&
+      !settings.has_device_id()) {
+    if (show_warnings()) {
+      std::cerr << "Warning: No GPU available for execution, falling back to"
+                   " using the host!"
+                << std::endl;
+    }
+    acc_set_device_type(acc_device_host);
+    Impl::OpenACCInternal::m_acc_device_num =
+        acc_get_device_num(acc_device_host);
+  } else {
+    using Kokkos::Impl::get_gpu;
+    int const dev_num = get_gpu(settings);
+    acc_set_device_num(dev_num, Impl::OpenACC_Traits::dev_type);
+    Impl::OpenACCInternal::m_acc_device_num = dev_num;
+  }
```

Comment:
One issue on moving the host-fallback handling code from `get_device_count()` in `Kokkos_Core.cpp` to here is that even if the host fallback is allowed, the Kokkos device initialization phase can fail if there exists any non-host device since `get_device_count()` will return 0. 
I think that a better way is to create a separate `detect_device_count()` function in the OpenACC backend, which handles the host fallback case, and `get_device_count()` internally calls the `detect_device_count()`. In this case, `get_device_count()` will always return correct device count.

Reply:
That code path you refer to would never be called.  It is exclusively called in `Kokkos::Impl::get_gpu`.

Reply:
The problem will occur if `Kokkos::Impl:get_gpu` is called at least once in such a case.
As I know, `Kokkos::Impl:get_gpu` is called only for device initialization, but if it is used in any future implementation, this latent problem will occur.

Reply:
I don't anticipate it will change and if we ever did that would break in other backends as well.

Reply:
(I don't know why but my last comment was deleted, and thus write again here)
The current implementation will be OK if we use `get_device_count` and `get_gpu` APIs only to refer to non-host devices (e.g., GPUs). Then, it will be OK for `get_device_count` to return 0 when there is no available non-host device, even if the host fallback is allowed (as long as `get_gpu` is not used at all in this case).